File: 프로그래머스/level1_크레인 인형뽑기 게임.py
from collections import deque

def solution(board, moves):
    q = deque(moves)
    bucket = []
    answer = 0
    while q:
        now = q.popleft() - 1

        for i in range(len(board)):
            if board[i][now] == 0: continue
            else:
                bucket.append(board[i][now])
                board[i][now] = 0
                break

        if len(bucket) >= 2:
            if bucket[-1] == bucket[-2]:
                answer += 2
                bucket.pop()
                bucket.pop()
    return answer


board = [[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]]
moves = [1,5,3,5,1,2,1,4]

print(solution(board, moves))

File: 프로그래머스 복습(java)/src/level1/level1_크레인_인형뽑기_게임.java
package level1;

import java.util.Stack;

public class level1_크레인_인형뽑기_게임 {
    static class Solution {
        public boolean isPop(Stack<Integer> stack, int x, int answer) {
            if (stack.isEmpty() || stack.peek() != x) {
                stack.push(x);
                return false;
            } else {
                stack.pop();
                return true;
            }
        }

        public int solution(int[][] board, int[] moves) {
            int answer = 0;
            Stack<Integer> stack = new Stack<>();
            int size = board.length;
            for (int i = 0; i < moves.length; i++) {
                int pick = moves[i] - 1;
                for (int j = 0; j < size; j++) {
                    if (board[j][pick] != 0) {
                        if (isPop(stack, board[j][pick], answer)) {
                            answer += 2;
                        }
                        board[j][pick] = 0;
                        break;
                    }
                }
            }
            return answer;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int board[][] = {{0, 0, 0, 0, 0}, {0, 0, 1, 0, 3}, {0, 2, 5, 0, 1}, {4, 2, 4, 4, 2}, {3, 5, 1, 3, 1}};
        int moves[] = {1, 5, 3, 5, 1, 2, 1, 4};
        System.out.println(solution.solution(board, moves));
    }
}
